### PR TITLE
#163383179 Add functionality for user to retrieve and edit their profile

### PR DESCRIPTION
**Pull request by oma0256**
**What does this PR do?**
 - Create a user's profile on successful user registration
 - Enable a user view their profile
 - Enable a user edit their profile
 - Allow a user view other users profiles
 - Prevent a user from editing other users profiles

**How should this be manually tested?**
 - Download and install postman onto you're local machine
 - Clone the repo onto you're local machine
 - Create and activate virtual environment
 - Move into the project directory
 `
pip3 install -r requirements.txt
python3 manage.py runserver
`
- Open postman
- Register a user using the route 
`http://127.0.0.1:8000/api/v1/users/`
- Login the user you just registered
`http://127.0.0.1:8000/api/v1/users/login/`
- Get the token that was returned
- Insert a header in postman with a key of `Authorization`  and value `Bearer <token>`
- Then visit the route `http://127.0.0.1:8000/api/v1/articles` to create an article using `POST`.
- Then visit the route `http://127.0.0.1:8000/api/v1/articles/<slug>/favorite` to either favourite or unfavourite an article `GET`.

**What are the relevant pivotal tracker stories?**
[#163383187](https://www.pivotaltracker.com/n/projects/2240000/stories/163383187)

### CODE REVIEW

**Review by fahdjamy — APPROVED** [review]

**Review by missvicki — COMMENTED** [review]
Oma nice work. Do we really have to use regular expressions in the url. Can't we avoid them?

**Comment by fahdjamy** [comment]
> Oma nice work. Do we really have to use regular expressions in the URL? Can't we avoid them?

Yeah, I think like we had agreed @oma0256. you should stick to using regular expressions in the urls.

**Comment by oma0256** [comment]
@missvicki had to use them in this case since there's no way I can match username without using regular expressions.

**Comment by codjoero** [comment]
@oma0256, Great work, but we are going to have to try and get rid of re_path see what happens.

**Comment by oma0256** [comment]
@codjoero @missvicki thanks for the feedback got rid of the regular expressions and used path

**Review by codjoero — APPROVED** [review]
I knew you could  do it, 
LGTM

**Comment by pbkabali** [comment]
Works fine. Good work @oma0256. I need to find you and you explain to me why we absolutely need the re_path in the url.

**Review by pbkabali — APPROVED** [review]

**Comment by oma0256** [comment]
@pbkabali It was a mistake on my side fixed it now using path instead of re_path

File: authors/apps/profiles/tests/test_update_profile.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+class TestUpdateProfile(BaseTest):
+    def test_patch_profile_successfully(self):
+        response = self.client.put(self.url,
+                                   data=json.dumps(valid_partial_profile_data),
+                                   content_type='application/json',
+                                   HTTP_AUTHORIZATION=f"Bearer {self.token}")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIsNotNone(response.data["bio"])
+
+    def test_update_profile_successfully(self):
```
Comment (karuhanga) [inline authors/apps/profiles/tests/test_update_profile.py:20]: I don't think this test achieves what you are trying to, that is you only check for whether or not what is returned is `None`, and not necessarily if it is the same as the update data.

**Review by karuhanga — CHANGES_REQUESTED** [review]
Good work on implementing this feature @oma0256.
A few thoughts;
- I think the names of the test methods could improve, for example it would be clearer to write `test_unauthenticated_user_cannot_update_a_profile` than `test_update_profile_when_unauthorized`, as the former clearly states what you need to achieve, or `test_user_cannot_update_another_users_profile` than `test_update_profile_for_different_user`.
- It is not clear from the updates to the `settings` how the image upload will work in production, especially given heroku's model does not support this. Could we perhaps we could look into uploading the images to another service like `aws` in production?

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -139,3 +141,12 @@ def token_generator(self):
         return token.decode('utf-8')
 
 
+def profile_post_save_reciever(sender, instance, created, *args, **kwargs):
```
Comment (malep2007) [inline authors/apps/authentication/models.py]: I think you should take advantage of python's kwargs to reduce the number of parameters that you pass to this method.
For this, you could easily re-write it as follows
```python
def profile_post_save_receiver(*args, **kwargs):
    sender = kwargs.get('sender')
    instance = kwargs.get('instance')
    created = kwargs.get('created')
    user = instance.user
    user.username = instance.username
    user.save(update_fields=['username])
```

Comment (malep2007) [reply authors/apps/authentication/models.py]: You can do the same for areas where you have many parameters in a method call. PEP8 requires no more than 4 parameters in a method call

File: authors/apps/profiles/tests/test_update_profile.py (hunk 1 above)
Comment (malep2007) [reply authors/apps/profiles/tests/test_update_profile.py:20]: I agree you should check to explicitly see of the updated `bio` information is what you expect. Remember tests should be explicit.

**Review by malep2007 — CHANGES_REQUESTED** [review]

**Review by malep2007 — APPROVED** [review]

**Pull request merged**